src/borsh.ts

```diff
@@ -0,0 +1,447 @@
+            return reader.readFixedArray(23);
+        }
+        case 'bytes24': {
+            return reader.readFixedArray(24);
+        }
+        case 'bytes25': {
+            return reader.readFixedArray(25);
+        }
+        case 'bytes26': {
+            return reader.readFixedArray(26);
+        }
+        case 'bytes27': {
+            return reader.readFixedArray(27);
+        }
+        case 'bytes28': {
+            return reader.readFixedArray(28);
+        }
+        case 'bytes29': {
+            return reader.readFixedArray(29);
+        }
+        case 'bytes30': {
+            return reader.readFixedArray(30);
+        }
+        case 'bytes31': {
+            return reader.readFixedArray(31);
+        }
+        case 'address':
+        case 'bytes32': {
+            return reader.readFixedArray(32);
+        }
+
+        case 'bool': {
+            return reader.readBool();
+        }
+
+        case 'bytes': {
+            const len = reader.readU32();
+            return reader.readFixedArray(len);
+        }
+
+        case 'string': {
+            return reader.readString();
+        }
+
+        case 'tuple': {
+            const tuple_len = param.components.length;
+            const response: Array<any> = new Array<any>();
+            for (let i = 0; i < tuple_len; i++) {
+                response.push(decodeParam(param.components[i], reader));
+            }
+            return response;
+        }
+
+        case 'array': {
+            let len = 0;
+            if (param.arrayLength == -1) {
+                len = reader.readU32();
+            } else {
+                len = param.arrayLength;
+            }
+            const response: Array<any> = new Array<any>();
+            for (let i = 0; i < len; i++) {
+                response.push(decodeParam(param.arrayChildren, reader));
+            }
+            return response;
+        }
+    }
+}
+
+/**
+ * This class extends BinaryReader to implement function for reading signed numbers from a borsh encoded buffer.
+ */
+class SolidityBinaryReader extends BinaryReader {
+    public constructor(buf: Buffer) {
+        super(buf);
+    }
+
+    readI8(): number {
+        const value = this.buf.getInt8(this.offset);
+        this.offset += 1;
+        return value;
+    }
+
+    readI16(): number {
+        const value = this.buf.getInt16(this.offset, true);
+        this.offset += 2;
+        return value;
+    }
+
+    readI32(): number {
+        const value = this.buf.getInt32(this.offset, true);
+        this.offset += 4;
+        return value;
+    }
+
+    readI64(): bigint {
+        const value = this.buf.getBigInt64(this.offset, true);
+        this.offset += 8;
+        return value;
+    }
+
+    readI128(): bigint {
+        const read = this.readFixedArray(16);
+        let bigNumber = new BN(read, 'le');
+        bigNumber = bigNumber.fromTwos(128);
+        const str = bigNumber.toString();
+        return BigInt(str);
```

Comment:
I can't find any direct way of creating a `BigInt` from some bytes either

Reply:
These lines are so inefficient code that I feel nausea when I read them.